imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,483 @@
+    existing_grids: typing.Optional[tuple[NDArray, NDArray]] = None,
+    # Only used if no existing grids are provided
+    spacing_deg: float = DEFAULT_SPACING_DEG,
+) -> tuple[NDArray]:
+    """
+    Project an az/el grid in an inertial frame (ECLIPJ2000) to the DPS frame.
+
+    # TODO: Likely replace J2000 with HAE frame (everywhere it occurs)
+
+    Parameters
+    ----------
+    event_time : float
+        Time at which to project the grid.
+    existing_grids : tuple[NDArray, NDArray], optional
+        Existing azimuth_array, elevation_array grids to use, by default None
+        in which case the grids will be built from scratch.
+        Each of these grids should be 2D arrays of either azimuth or elevation values.
+    spacing_deg : float, optional
+        Spacing of the grid in degrees, by default DEFAULT_SPACING_DEG.
+
+    Returns
+    -------
+    tuple[NDArray]
+        Tuple of following arrays and dimensions:
+        - Azimuth values of all points on inertial grid (2D)
+        - Elevation grid of all points on inertial grid (2D)
+        - Flat indices of these grid points in the heliocentric frame (1D)
+        - Flat indices of these grid points projected into the DPS frame (1D)
+        - Azimuth values of original points in the inertial frame frame,
+        raveled and projected into DPS azimuth coords (1D)
+        - Elevation values of original points in the inertial frame frame,
+        raveled and projected into DPS elevation coords (1D)
+        - Azimuth indices of original points in the DPS azimuth grid (1D)
+        - Elevation indices of original points in the DPS elevation grid (1D)
+    """
+    if not existing_grids:
+        # Build the azimuth, elevation grid in the inertial frame
+        (_, __, az_grid, el_grid) = l2_utils.build_az_el_grid(
+            spacing=spacing_deg,
+            input_degrees=True,
+            output_degrees=False,
+            centered_azimuth=False,  # (0, 2pi rad = 0, 360 deg)
+            centered_elevation=True,  # (-pi/2, pi/2 rad = -90, 90 deg)
+        )
+    else:
+        az_grid, el_grid = existing_grids
+
+    # Unwrap the grid to a 1D array for same interface as tessellation
+    az_grid_raveled = az_grid.ravel(order="F")
+    el_grid_raveled = el_grid.ravel(order="F")
+
+    radii_helio = geometry.spherical_to_cartesian(
+        np.stack(
+            (np.ones_like(az_grid_raveled), az_grid_raveled, el_grid_raveled),
+            axis=-1,
+        )
+    )
+    flat_indices_helio = np.arange(az_grid_raveled.size)
+
+    # TODO: reset the frame definition for the appropriate Pointing
+    # I'm not sure where to get the various spice kernels
+
+    rotation_matrix_helio_to_dps = spice.pxform(
+        fromstr="ECLIPJ2000", tostr="IMAP_DPS", et=event_time
+    )
+
+    # cartesian (x,y,z) radii vectors in the DPS frame corresponding to
+    # the grid points in the helio inertial frame
+    radii_dps = np.dot(rotation_matrix_helio_to_dps, radii_helio.T).T
```

Comment:
This can be done with an existing function in the spice.geometry module:
```suggestion
    from imap_processing.spice.geometry import frame_transform, SpiceFrame
    radii_dps = frame_transform(event_time, radii_helio, SpiceFrame.ECLIPJ2000, SpiceFrame.IMAP_DPS)
```
With this change, it looks like this function will no longer need the `@ensure_spice` decorator.

Reply:
Ah! That's helpful, thanks!

Reply:
@subagonsouth - I don't actually know if I can use `frame_transform`, since it will only apply to a single vector for each et. I could pass it a whole list of all the same et, but then I think it would re-calculate the same rotation matrix n-times for each of the $360 * 180 / (spacing^2)$ vectors. 

How about I just create the rotation matrix using our geometry.get_rotation_matrix function, and apply the rotation matrix in the same way?

Reply:
Another option is altering `frame_transform` to allow multiple vectors for the same et. Given that the function explicitly forbids that, I figured there was a reason not to do so?

Reply:
I think that altering `frame_transform` is a good idea. I think that it was my failure to be creative enough to think of that use case that caused that situation to be disallowed.

Reply:
![Screenshot 2024-12-05 at 3 14 51 PM](https://github.com/user-attachments/assets/c51cf3ee-e428-406c-b0ff-a18ce67dd0a5)


I'm having it still raise the error, unless the len(et) == 1, because I think that's the only case where we can safely do this (e.g. if there are m `position`s and n `et`s, that's no good). I think that would fail within numpy broadcasting anyways, but no 100%.